src/Features/Core/Portable/NavigateTo/AbstractNavigateToSearchService.Remote.cs

```diff
@@ -1,38 +1,44 @@
 ﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
 
+using System;
 using System.Collections.Immutable;
+using System.Linq;
 using System.Threading;
 using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Experiments;
 using Microsoft.CodeAnalysis.Remote;
 
 namespace Microsoft.CodeAnalysis.NavigateTo
 {
     internal abstract partial class AbstractNavigateToSearchService
     {
         private async Task<ImmutableArray<INavigateToSearchResult>> SearchDocumentInRemoteProcessAsync(
-            RemoteHostClient.Session session, Document document, string searchPattern, CancellationToken cancellationToken)
+            RemoteHostClient client, Document document, string searchPattern, CancellationToken cancellationToken)
         {
-            var serializableResults = await session.InvokeAsync<ImmutableArray<SerializableNavigateToSearchResult>>(
-                nameof(IRemoteNavigateToSearchService.SearchDocumentAsync),
-                document.Id, searchPattern).ConfigureAwait(false);
+            var solution = document.Project.Solution;
 
-            return serializableResults.SelectAsArray(r => r.Rehydrate(document.Project.Solution));
+            var serializableResults = await client.RunCodeAnalysisServiceOnRemoteHostAsync<ImmutableArray<SerializableNavigateToSearchResult>>(
+                solution, nameof(IRemoteNavigateToSearchService.SearchDocumentAsync),
+                new object[] { document.Id, searchPattern }, cancellationToken).ConfigureAwait(false);
+
+            return serializableResults.NullToEmpty().SelectAsArray(r => r.Rehydrate(solution));
```

Comment:
ya, I should pick one pattern and stick to it. I made it to return tuple since you or Jason I am not sure, but one of you left feedback before saying returning default on failure is error prone and want API to return success or failure explicitly. since async method can't have out parameter, I made it to return tuple.